File: lab/javascript/textos/contar_string.js
function contarPalavras(texto) {
    // Normaliza: deixa em minúsculas e divide por espaço
    let palavras = texto.toLowerCase().split(/\s+/);

    let freq = {};

    for (let palavra of palavras) {
        if (/^[a-zá-ú]+$/i.test(palavra)) { // só letras
            freq[palavra] = (freq[palavra] || 0) + 1;
        }
    }

    // Converte para array de [palavra, contagem] e ordena
    let ordenado = Object.entries(freq).sort((a, b) => b[1] - a[1]);

    // Exibe
    for (let [palavra, qtd] of ordenado) {
        console.log(`${palavra}: ${qtd}`);
    }
}


function contarPalavras2(texto) {
    // deixa tudo minúsculo e separa por espaço
    let palavras = texto.toLowerCase().split(" ");
    let mapa = {};

    // conta ocorrências
    for (let palavra of palavras) {
        if (/^[a-záéíóúãõâêîôûç]+$/i.test(palavra)) { // só letras
            mapa[palavra] = (mapa[palavra] || 0) + 1;
        }
    }

    // transforma em lista de [palavra, freq]
    let lista = Object.entries(mapa);

    // selection sort pela frequência (decrescente)
    for (let i = 0; i < lista.length; i++) {
        let maxIdx = i;
        for (let j = i + 1; j < lista.length; j++) {
            if (lista[j][1] > lista[maxIdx][1]) {
                maxIdx = j;
            }
        }
        // swap
        [lista[i], lista[maxIdx]] = [lista[maxIdx], lista[i]];
    }

    // imprime resultado
    for (let [palavra, freq] of lista) {
        console.log(`${palavra}: ${freq}`);
    }
}


// Exemplo
contarPalavras("hoje é dia de estudar estudar estudar python python");
console.log("----");
contarPalavras2("hoje é dia de estudar estudar estudar python python");

File: lab/c/textos/contar_string.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define MAX_PALAVRAS 100
#define MAX_LEN 50

typedef struct {
    char palavra[MAX_LEN];
    int contagem;
} PalavraFreq;

int main() {
    char texto[] = "hoje e dia de estudar estudar estudar python python";
    char *token;
    PalavraFreq freq[MAX_PALAVRAS];
    int total = 0;

    // Tokenizar por espaço
    token = strtok(texto, " ");
    while (token != NULL) {
        // Normalizar para minúsculo
        for (int i = 0; token[i]; i++)
            token[i] = tolower(token[i]);

        // Procurar se já existe
        int encontrado = 0;
        for (int i = 0; i < total; i++) {
            if (strcmp(freq[i].palavra, token) == 0) {
                freq[i].contagem++;
                encontrado = 1;
                break;
            }
        }
        // Se não existe, adiciona
        if (!encontrado) {
            strcpy(freq[total].palavra, token);
            freq[total].contagem = 1;
            total++;
        }

        token = strtok(NULL, " ");
    }

    // Ordenar por contagem (bubble sort simples)
    for (int i = 0; i < total - 1; i++) {
        for (int j = i + 1; j < total; j++) {
            if (freq[j].contagem > freq[i].contagem) {
                PalavraFreq temp = freq[i];
                freq[i] = freq[j];
                freq[j] = temp;
            }
        }
    }

    // Exibir
    for (int i = 0; i < total; i++) {
        printf("%s: %d\n", freq[i].palavra, freq[i].contagem);
    }

    return 0;
}
